#

## Import thư viện cần thiết

In [106]:
import pandas as pd
import numpy as np
import csv
import matplotlib.pyplot as plt

## Load dataset superstore

In [107]:
df=pd.read_csv('../data/SuperStoreOrders - SuperStoreOrders.csv', encoding='utf-8')

In [108]:
df.head()

,order_id,order_date,ship_date,ship_mode,customer_name,segment,state,country,market,region,...,category,sub_category,product_name,sales,quantity,discount,profit,shipping_cost,order_priority,year
0,AG-2011-2040,1/1/2011,6/1/2011,Standard Class,Toby Braunhardt,Consumer,Constantine,Algeria,Africa,Africa,...,Office Supplies,Storage,"Tenex Lockers, Blue",408,2,0.0,106.140,35.46,Medium,2011
1,IN-2011-47883,1/1/2011,8/1/2011,Standard Class,Joseph Holt,Consumer,New South Wales,Australia,APAC,Oceania,...,Office Supplies,Supplies,"Acme Trimmer, High Speed",120,3,0.1,36.036,9.72,Medium,2011
2,HU-2011-1220,1/1/2011,5/1/2011,Second Class,Annie Thurman,Consumer,Budapest,Hungary,EMEA,EMEA,...,Office Supplies,Storage,"Tenex Box, Single Width",66,4,0.0,29.640,8.17,High,2011
3,IT-2011-3647632,1/1/2011,5/1/2011,Second Class,Eugene Moren,Home Office,Stockholm,Sweden,EU,North,...,Office Supplies,Paper,"Enermax Note Cards, Premium",45,3,0.5,-26.055,4.82,High,2011
4,IN-2011-47883,1/1/2011,8/1/2011,Standard Class,Joseph Holt,Consumer,New South Wales,Australia,APAC,Oceania,...,Furniture,Furnishings,"Eldon Light Bulb, Duo Pack",114,5,0.1,37.770,4.70,Medium,2011


In [109]:
print(f"Dataset có số hàng: {df.shape[0]} và số cột: {df.shape[1]}")

Dataset có số hàng: 51290 và số cột: 21


In [110]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51290 entries, 0 to 51289
Data columns (total 21 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   order_id        51290 non-null  object 
 1   order_date      51290 non-null  object 
 2   ship_date       51290 non-null  object 
 3   ship_mode       51290 non-null  object 
 4   customer_name   51290 non-null  object 
 5   segment         51290 non-null  object 
 6   state           51290 non-null  object 
 7   country         51290 non-null  object 
 8   market          51290 non-null  object 
 9   region          51290 non-null  object 
 10  product_id      51290 non-null  object 
 11  category        51290 non-null  object 
 12  sub_category    51290 non-null  object 
 13  product_name    51290 non-null  object 
 14  sales           51290 non-null  object 
 15  quantity        51290 non-null  int64  
 16  discount        51290 non-null  float64
 17  profit          51290 non-null 

In [111]:
df.isnull().sum()

order_id          0
order_date        0
ship_date         0
ship_mode         0
customer_name     0
segment           0
state             0
country           0
market            0
region            0
product_id        0
category          0
sub_category      0
product_name      0
sales             0
quantity          0
discount          0
profit            0
shipping_cost     0
order_priority    0
year              0
dtype: int64

dataset của chúng ta có 51290 rows và 21 features và mỗi feature là đều có đầy đủ 51290 giá trị và  non-null nên -> Không có missing value

In [112]:
df.describe()

,quantity,discount,profit,shipping_cost,year
count,51290.000000,51290.000000,51290.000000,51290.000000,51290.000000
mean,3.476545,0.142908,28.641740,26.375915,2012.777208
std,2.278766,0.212280,174.424113,57.296804,1.098931
min,1.000000,0.000000,-6599.978000,0.000000,2011.000000
25%,2.000000,0.000000,0.000000,2.610000,2012.000000
50%,3.000000,0.000000,9.240000,7.790000,2013.000000
75%,5.000000,0.200000,36.810000,24.450000,2014.000000
max,14.000000,0.850000,8399.976000,933.570000,2014.000000


In [113]:
print(df.dtypes)


order_id           object
order_date         object
ship_date          object
ship_mode          object
customer_name      object
segment            object
state              object
country            object
market             object
region             object
product_id         object
category           object
sub_category       object
product_name       object
sales              object
quantity            int64
discount          float64
profit            float64
shipping_cost     float64
order_priority     object
year                int64
dtype: object


Ta thấy sales đang là object nhưng thực tế phải là float và order_date,ship_date  phải là datetime thay vì object

In [114]:
clean_df = df.copy()


In [115]:
clean_df['sales']=clean_df['sales'].astype(str).str.replace('$','').str.replace(',','')


In [116]:
for col in ['sales', 'quantity', 'discount', 'profit', 'shipping_cost']:
    clean_df[col] = pd.to_numeric(clean_df[col], errors='coerce')

In [117]:
clean_df['order_date'] = pd.to_datetime(clean_df['order_date'], dayfirst=True)
clean_df['ship_date']  = pd.to_datetime(clean_df['ship_date'],  dayfirst=True)

In [118]:
clean_df.isnull().sum()

order_id          0
order_date        0
ship_date         0
ship_mode         0
customer_name     0
segment           0
state             0
country           0
market            0
region            0
product_id        0
category          0
sub_category      0
product_name      0
sales             0
quantity          0
discount          0
profit            0
shipping_cost     0
order_priority    0
year              0
dtype: int64

In [119]:
clean_df.dtypes

order_id                  object
order_date        datetime64[ns]
ship_date         datetime64[ns]
ship_mode                 object
customer_name             object
segment                   object
state                     object
country                   object
market                    object
region                    object
product_id                object
category                  object
sub_category              object
product_name              object
sales                      int64
quantity                   int64
discount                 float64
profit                   float64
shipping_cost            float64
order_priority            object
year                       int64
dtype: object

In [120]:
n_cols=clean_df.select_dtypes(include=['number']).columns
print("Các cột có kiểu dữ liệu số:")
for col in n_cols:
    print(f"{col}")

Các cột có kiểu dữ liệu số:
sales
quantity
discount
profit
shipping_cost
year


In [121]:
o_cols=clean_df.select_dtypes(include=['object']).columns
print("Các cột có kiểu dữ liệu object:")
for col in o_cols:
    print(f"{col}")
    

Các cột có kiểu dữ liệu object:
order_id
ship_mode
customer_name
segment
state
country
market
region
product_id
category
sub_category
product_name
order_priority


In [122]:
# Loại bỏ khoảng trắng thừa ở đầu và cuối chuỗi trong các cột object
for col in o_cols:
    clean_df[col]=clean_df[col].str.strip()

In [123]:
# Kiểm tra và loại bỏ các dòng trùng lặp
print("Các dòng bị trùng lặp : ",clean_df.duplicated().sum())
clean_df=clean_df.drop_duplicates()

Các dòng bị trùng lặp :  0


In [124]:
# Đảm bảo rằng ngày giao hàng không sớm hơn ngày đặt hàng
clean_df[(clean_df['ship_date'] < clean_df['order_date'])]

,order_id,order_date,ship_date,ship_mode,customer_name,segment,state,country,market,region,...,category,sub_category,product_name,sales,quantity,discount,profit,shipping_cost,order_priority,year


In [125]:
(clean_df['ship_date'] < clean_df['order_date']).sum()

np.int64(0)

In [126]:
clean_df['segment'].value_counts()

segment
Consumer       26518
Corporate      15429
Home Office     9343
Name: count, dtype: int64

In [127]:
clean_df.loc[((clean_df['sales'] !=0) & (clean_df['profit'] < 0)),['profit','sales']].sort_values(by='profit')

,profit,sales
31884,-6599.9780,4500
29212,-4088.3760,3085
47195,-3839.9904,8000
3775,-3701.8928,2178
37130,-3399.9800,2550
...,...,...
26048,-0.0060,271
15559,-0.0040,156
33388,-0.0030,159
19835,-0.0021,562


In [128]:

(clean_df['sales']<clean_df['profit']).sum()

np.int64(0)

In [129]:
# Tạo cột profit_margin để tính tỷ suất lợi nhuận
clean_df['profit_margin']=np.where(clean_df['sales'] != 0, clean_df['profit'] / clean_df['sales'],0).round(4)


In [130]:
# Tạo cột is_loss để đánh dấu các đơn hàng có lợi nhuận âm
clean_df['is_loss']=(clean_df['profit'] < 0).astype(int)

In [131]:
def discount_level(discount):
    if discount == 0:
        return 'No'
    elif discount < 0.2:
        return 'Low'
    elif discount < 0.4:
        return 'Medium'
    else:
        return 'High'

In [132]:
# Tạo cột discount_level để phân loại mức độ giảm giá ở đây sẽ có 4 mức: No, Low, Medium, High
clean_df['discount_level']=clean_df['discount'].apply(discount_level)
clean_df['discount_level'].value_counts()


discount_level
No        29009
High      10138
Medium     6188
Low        5955
Name: count, dtype: int64

In [133]:
clean_df['shipping_days']  = (clean_df['ship_date'] - clean_df['order_date']).dt.days


## Validation

In [134]:
print(len(clean_df) == len(df))


True


In [135]:
print(clean_df.isnull().sum())


order_id          0
order_date        0
ship_date         0
ship_mode         0
customer_name     0
segment           0
state             0
country           0
market            0
region            0
product_id        0
category          0
sub_category      0
product_name      0
sales             0
quantity          0
discount          0
profit            0
shipping_cost     0
order_priority    0
year              0
profit_margin     0
is_loss           0
discount_level    0
shipping_days     0
dtype: int64


In [136]:
clean_df.duplicated().sum()

np.int64(0)

In [137]:
clean_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51290 entries, 0 to 51289
Data columns (total 25 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   order_id        51290 non-null  object        
 1   order_date      51290 non-null  datetime64[ns]
 2   ship_date       51290 non-null  datetime64[ns]
 3   ship_mode       51290 non-null  object        
 4   customer_name   51290 non-null  object        
 5   segment         51290 non-null  object        
 6   state           51290 non-null  object        
 7   country         51290 non-null  object        
 8   market          51290 non-null  object        
 9   region          51290 non-null  object        
 10  product_id      51290 non-null  object        
 11  category        51290 non-null  object        
 12  sub_category    51290 non-null  object        
 13  product_name    51290 non-null  object        
 14  sales           51290 non-null  int64         
 15  qu

## Xuất file đã cleaning

In [138]:

output = '../data/clean.csv'
clean_df.to_csv(output, index=False,encoding='utf-8-sig', quoting=2)
print(f' {output}: {len(clean_df)} rows, {clean_df.shape[1]} cols')

 ../data/clean.csv: 51290 rows, 25 cols
